File: acmp/arch/3.Динамическое-программирование/0.Лекции/01.Задача о сдаче/main.py


# read input
s = open("INPUT.TXT").read().split();

toklen = int(s[0]);
tokens = list();

for i in range(0, toklen):
	tokens.append( int(s[i+1]) );

n = int(s[toklen+1]);


# calculate
variants = list();
variants.append(1);

for i in range(1, n+1):
	variants.append(0);
	for j in range(0, toklen):
		if tokens[j] > i:
			break;
		variants[i] += variants[ i-tokens[j] ];

print( variants[ len(variants)-1 ] );

File: acmp/arch/3.Динамическое-программирование/0.Лекции/01.Задача о сдаче/main.cpp
#include <algorithm>
#include <iostream>
#include <fstream>


using namespace std;





char const *INPUT_FILE_NAME = "INPUT.TXT";
char const *OUTPUT_FILE_NAME = "OUTPUT.TXT";


int toklen;
int *tokens;
int n;
int *vars;



void calculate_variants()
{
	vars[0] = 1;
	for(int i = 1; i < n+1; ++i)
	{
		vars[i] = 0;
		for(
			auto *b = tokens, *e = tokens+toklen;
			b != e && *b <= i; ++b
		)
		{
			vars[i] += vars[i-*b];
		}
	}
}




int main( int argc, char *argv[] )
{
	{
		ifstream fin(INPUT_FILE_NAME);

		// tokens
		fin >> toklen;
		tokens = new int[toklen];
		for(auto *b = tokens, *e = tokens+toklen; b != e; ++b)
		{
			fin >> *b;
		}
		sort(tokens, tokens+toklen);

		// n, vars
		fin >> n;
		vars = new int[n+1];
	}

	calculate_variants();

	cout << vars[n] << endl;
	

	return 0;
}





// end
